.circleci/config.yml

```diff
@@ -0,0 +1,159 @@
+version: 2.1
+
+defaults: &defaults
+  working_directory: ~/repo
+  environment:
+    LEIN_ROOT: "true"
+    JVM_OPTS: -Xmx3200m
+
+
+executors:
+  openjdk8:
+    docker:
+      - image: circleci/clojure:openjdk-8-lein-2.9.0
+    <<: *defaults
+  openjdk11:
+    docker:
+      - image: circleci/clojure:openjdk-11-lein-2.9.0
+    <<: *defaults
+
+jobs:
+  checkout_code:
+    executor: openjdk11
+    steps:
+      - checkout
+      - restore_cache:
+          keys:
+            - v2-dependencies-{{ checksum "project.clj" }}
+            - v2-dependencies-
+      - run:
+          name: Downloading deps
+          command: |
+            lein deps
+      - save_cache:
+          paths:
+            - ~/.m2
+          key: v2-dependencies-{{ checksum "project.clj" }}
+      - persist_to_workspace:
+          root: /home/circleci
+          paths:
+            - .m2
+            - repo
+
+  lint_code:
+    executor: openjdk11
+    steps:
+      - attach_workspace:
+          at: /home/circleci
+      - run:
+          name: Install make
+          command: |
+            sudo apt-get install make
+      - run:
+          name: Running Eastwood
+          command: |
+            make eastwood
+      - run:
+          name: Running cljfmt
+          command: |
+            make cljfmt
+
+  test_code:
+    description: Run tests
+    parameters:
+      jdk_version:
+        description: Version of JDK to test against
+        type: string
+      clojure_version:
+        description: Version of Clojure to test against
+        type: string
+    executor: << parameters.jdk_version >>
+    environment:
+      VERSION: << parameters.clojure_version >>
+    steps:
+      - attach_workspace:
+          at: /home/circleci
+      - run:
+          name: Install make
+          command: |
+            sudo apt-get install make
+      - run:
+          name: Running tests
+          command: make test
+
+  codecov:
+    executor: openjdk11
+    environment:
+      VERSION: "1.9"
+    steps:
+      - attach_workspace:
+          at: /home/circleci
+      - run:
+          name: Install make
+          command: |
+            sudo apt-get install make
+      - run:
+          name: Running cloverage
+          command: |
+            make cloverage
+
+workflows:
+  version: 2.1
+  build_and_test:
+    jobs:
+      - checkout_code
+      - lint_code:
+          name: Code Linting
+          requires:
```

Comment:
What's with all those requires? Isn't there some way to avoid this repetition?

Reply:
The gist of this is that we check out the code + download deps once, and test n times. Thus each of the tests require the `checkout_code` step.

Alternatively, we can just run n tests, and have each job do its own checkout etc. I'm not sure either is better than the other given that everything runs fast enough.

Answering the next question: in CircleCI, each you you commit, you trigger a workflow, which is comprised of many jobs. You can set up dependency between the jobs, so they run in order while allowing for parallel jobs etc.